packages/core/bootstrap/src/lib/ws/epics.ts

```diff
@@ -0,0 +1,183 @@
+import { AnyAction } from 'redux'
+import { combineEpics, createEpicMiddleware, Epic } from 'redux-observable'
+import { merge, Subject } from 'rxjs'
+import {
+  endWith,
+  filter,
+  map,
+  mergeMap,
+  startWith,
+  takeUntil,
+  tap,
+  withLatestFrom,
+} from 'rxjs/operators'
+import { webSocket } from 'rxjs/webSocket'
+import WebSocket from 'ws'
+import {
+  connect,
+  connected,
+  disconnect,
+  disconnected,
+  subscribe,
+  subscribed,
+  unsubscribe,
+  unsubscribed,
+  messageReceived,
+  WSConfigPayload,
+  WSSubscriptionPayload,
+} from './actions'
+import {
+  ws_connection_active,
+  ws_message_total,
+  ws_subscription_active,
+  ws_subscription_total,
+} from './metrics'
+
+export const connectEpic: Epic<AnyAction, AnyAction, any, any> = (action$, state$) =>
+  action$.pipe(
+    filter(connect.match),
+    map(({ payload }) => ({ payload, connectionKey: payload.config.connectionInfo.key })),
+    // check if the connection already exists, then noop
+    withLatestFrom(state$),
+    filter(([{ connectionKey }, state]) => {
+      // if connection does not exist, then continue
+      return !state.ws.connections.active[connectionKey]
+    }),
+    // on a connect action being dispatched, open a new WS connection if one doesn't exist yet
+    mergeMap(([{ connectionKey, payload }]) => {
+      const { url, protocol } = payload.config.connectionInfo
+      const openObserver = new Subject()
+      const closeObserver = new Subject()
+      const WebSocketCtor = WebSocket
+      const wsSubject = webSocket({
+        url,
+        protocol,
+        openObserver,
+        closeObserver,
+        WebSocketCtor: WebSocketCtor as any, // TODO: fix types don't match
+      })
+
+      const { config } = payload
+      // Stream of WS connected & disconnected events
+      const open$ = openObserver.pipe(map((event) => connected({ event, config } as any)))
+      const close$ = closeObserver.pipe(map((event) => disconnected({ event, config } as any)))
+
+      // Close the WS connection on disconnect
+      const disconnect$ = action$.pipe(
+        filter(disconnect.match),
+        filter(({ payload }) => payload.connectionInfo.key === connectionKey),
+        tap(() => wsSubject.closed || wsSubject.complete()),
+        filter(() => false), // do not duplicate events
```

Comment:
I think this should be replaced with a no-op action. Such as `WS/UNSUBSCRIBED_SUCCESS`. Usually, with a single async action there is a minimum set of 2 actions, with 3 actions being the norm for error handling. E.g.

For unsubscribing, we would have the following set of actions:
`WS/UNSUBSCRIBE_REQUESTED`
`WS/UNSUBSCRIBE_SUCCESSFUL`
`WS/UNSUBSCRIBE_FAILED`

To reduce the amount of boilerplate this generates, we can use the `createSlice` helper function provided by redux toolkit which automatically creates actions based on the configured reducer keys

Reply:
@RodrigoAD please add another GitHub issue: Improve Redux command handling. 
I agree every command should be followed by a `SUCCESS` or `FAIL` event.

Reply:
#470